primitives/src/vstaging/executor_params.rs

```diff
@@ -0,0 +1,123 @@
+// Copyright 2021 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Abstract execution environment parameter set.
+//!
+//! Parameter set is encoded as an opaque vector which structure depends on the execution
+//! environment itself (except for environment type/version which is always represented
+//! by the first element of the vector). Decoding to a usable semantics structure is
+//! done in `polkadot-node-core-pvf`.
+
+use parity_scale_codec::{Decode, Encode};
+use parity_util_mem::MallocSizeOf;
+use scale_info::TypeInfo;
+use sp_std::{ops::Deref, vec::Vec};
+
+/// Execution environment type
+#[derive(Clone, Copy, Debug, Encode, Decode, PartialEq, Eq, MallocSizeOf, TypeInfo)]
+pub enum ExecutionEnvironment {
+	/// Generic Wasmtime executor
+	WasmtimeGeneric = 0,
+}
+
+/// Executor instantiation strategy
+#[derive(Clone, Debug, Encode, Decode, PartialEq, Eq, MallocSizeOf, TypeInfo)]
+pub enum ExecInstantiationStrategy {
+	/// Pooling copy-on-write
+	PoolingCoW,
+	/// Recreate instance copy-on-write
+	RecreateCoW,
+	/// Pooling
+	Pooling,
+	/// Recreate instance
+	Recreate,
+	/// Legacy instantiation strategy
+	Legacy,
+}
+
+/// A single executor parameter
+#[non_exhaustive]
+#[derive(Clone, Debug, Encode, Decode, PartialEq, Eq, MallocSizeOf, TypeInfo)]
+pub enum ExecutorParam {
+	/// General parameters:
+	/// Execution environment type and a version of the set of execution parameters, unique in
+	/// the scope of execution environment type
+	Version(ExecutionEnvironment, u32),
+
+	/// Parameters setting the executuion environment semantics:
+	/// Number of extra heap pages
+	ExtraHeapPages(u64),
+	/// Max. memory size
+	MaxMemorySize(u32),
+	/// Wasm logical stack size limit, in units
+	StackLogicalMax(u32),
+	/// Executor machine stack size limit, in bytes
+	StackNativeMax(u32),
+	/// Executor instantiation strategy
+	InstantiationStrategy(ExecInstantiationStrategy),
+	/// `true` if compiler should perform NaNs canonicalization
+	CanonicalizeNaNs(bool),
+	/// `true` if parallel compilation is allowed, single thread is used otherwise
+	ParallelCompilation(bool),
+}
```

Comment:
We discussed a different approach to this.

Specifically, my opinion was that we should **conceptually** use a mapping from a `u32` to a `Vec<u8>`. The main reason was that `enum` provides a minimal improvement on the developer UX, but at the same time raises some critical questions.

Let's examine what would happen if we were to add another variant to this enum. Once the runtime uses the new variant, what an old node client would do? Scale's `decode` would not be able to deserialize this data structure. 

This is not critical because the node client would have returned an error down the line anyway since the node won't recognize that field. 

Nonetheless, I still think that conceptually it's not correct to depend on a serialization error for such business logic.